free/src/main/scala/cats/free/FreeT.scala

```diff
@@ -0,0 +1,258 @@
+      case Suspend(m) =>
+        Suspend(M.map(m)(_.map(s => st(s))))
+    }
+
+  /**
+    * Runs to completion, mapping the suspension with the given transformation
+    * at each step and accumulating into the monad `M`.
+    */
+  def foldMap(f: S ~> M)(implicit MR: MonadRec[M]): M[A] = {
+    def go(ft: FreeT[S, M, A]): M[FreeT[S, M, A] Xor A] =
+      ft match {
+        case Suspend(ma) => MR.flatMap(ma) {
+          case Xor.Left(a) => MR.pure(Xor.Right(a))
+          case Xor.Right(sa) => MR.map(f(sa))(Xor.right)
+        }
+        case g @ FlatMapped(_, _) => g.a match {
+          case Suspend(mx) => MR.flatMap(mx) {
+            case Xor.Left(x) => MR.pure(Xor.left(g.f(x)))
+            case Xor.Right(sx) => MR.map(f(sx))(g.f andThen Xor.left)
+          }
+          case g0 @ FlatMapped(_, _) => MR.pure(Xor.left(g0.a.flatMap(g0.f(_).flatMap(g.f))))
+        }
+      }
+
+    MR.tailRecM(this)(go)
+  }
+
+  /** Evaluates a single layer of the free monad */
+  def resume(implicit S: Functor[S], MR: MonadRec[M]): M[A Xor S[FreeT[S, M, A]]] = {
+    def go(ft: FreeT[S, M, A]): M[FreeT[S, M, A] Xor (A Xor S[FreeT[S, M, A]])] =
+      ft match {
+        case Suspend(f) => MR.map(f)(as => Xor.right(as.map(S.map(_)(pure(_)))))
+        case g1 @ FlatMapped(_, _) => g1.a match {
+          case Suspend(m1) => MR.map(m1) {
+            case Xor.Left(a) => Xor.left(g1.f(a))
+            case Xor.Right(fc) => Xor.right(Xor.right(S.map(fc)(g1.f(_))))
+          }
+          case g2 @ FlatMapped(_, _) => MR.pure(Xor.left(g2.a.flatMap(g2.f(_).flatMap(g1.f))))
+        }
+      }
+
+    MR.tailRecM(this)(go)
+  }
+
+  /**
+    * Runs to completion, using a function that maps the resumption from `S` to a monad `M`.
+    */
+  def runM(interp: S[FreeT[S, M, A]] => M[FreeT[S, M, A]])(implicit S: Functor[S], MR: MonadRec[M]): M[A] = {
+    def runM2(ft: FreeT[S, M, A]): M[FreeT[S, M, A] Xor A] =
+      MR.flatMap(ft.resume) {
+        case Xor.Left(a) => MR.pure(Xor.right(a))
+        case Xor.Right(fc) => MR.map(interp(fc))(Xor.left)
+      }
+    MR.tailRecM(this)(runM2)
+  }
+
+  /**
+   * Finds the first `M` instance, `m`, and maps it to contain the rest
+   * of the computation. Since only `map` is used on `m`, its structure
+   * is preserved.
+   */
+  @tailrec
+  private[cats] final def toM(implicit M: Applicative[M]): M[FreeT[S, M, A]] =
+    this match {
+      case Suspend(m) => M.map(m) {
+        case Xor.Left(a) => pure(a)
+        case Xor.Right(s) => liftF(s)
+      }
+      case g1 @ FlatMapped(_, _) => g1.a match {
+        case Suspend(m) => M.map(m) {
+          case Xor.Left(a) => g1.f(a)
+          case Xor.Right(s) => liftF[S, M, g1.A](s).flatMap(g1.f)
+        }
+        case g0 @ FlatMapped(_, _) => g0.a.flatMap(g0.f(_).flatMap(g1.f)).toM
+      }
+    }
+
+  @tailrec
+  private def step: FreeT[S, M, A] =
+    this match {
+      case g @ FlatMapped(_, _) => g.a match {
```

Comment:
what's the rationale for `g @ FlatMapped(_, _) => g.a` rather than `FlatMapped(a, _) =>`

same below.

Also, why not:
```scala
this match {
  case FlatMapped(FlatMapped(a, fn0), fn1) =>
    a.flatMap(a => fn0(a).flatMap(fn1)).step
  case nonNestedFlatMap => nonNestedFlatMap
}
```